tfhe/src/core_crypto/gpu/vec.rs

```diff
@@ -66,6 +67,25 @@ impl<T: Numeric> CudaVec<T> {
     pub fn is_empty(&self) -> bool {
         self.len == 0
     }
+
+    pub(crate) fn range_bounds_to_start_end<R>(&self, range: R) -> std::ops::RangeInclusive<usize>
+    where
+        R: std::ops::RangeBounds<usize>,
+    {
+        let start = match range.start_bound() {
+            Unbounded => 0usize,
+            Included(start) => *start,
+            Excluded(start) => *start + 1,
+        };
+
+        let end = match range.end_bound() {
+            Unbounded => self.len() - 1,
+            Included(end) => *end,
+            Excluded(end) => *end - 1,
+        };
```

Comment:
underflows if self.len == 0, the sanity check above (reproduce here) will fail, consider using saturating sub

```rust
    pub unsafe fn copy_dest_range_gpu_to_gpu_async<R, T>(
        &self,
        range: R,
        dest: &mut CudaVec<T>,
        src: &CudaVec<T>,
    ) where
        R: std::ops::RangeBounds<usize>,
        T: Numeric,
    {
        let (start, end) = dest.range_bounds_to_start_end(range).into_inner();
        // size is > 0 thanks to this check
        if end < start 
```

Reply:
Thank you, I've used saturating_sub for the two subtractions in `range_bounds_to_start_end`.